File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int dive(TreeNode* node, int depth) {
        if(!node) {
            return depth - 1;
        }
        return max(dive(node->left, depth + 1), dive(node->right, depth + 1));
    }

    int maxDepth(TreeNode* root) {
        if (!root) {
            return 0;
        }
        return dive(root, 1);
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.max_diameter = 0

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        def dfs(node: Optional[TreeNode]) -> int:
            if node is None:
                return 0
            else:
                leftHeight = dfs(node.left)
                rightHeight = dfs(node.right)
                height_node = 1 + max(leftHeight, rightHeight)
                print("height node ", height_node)
                self.max_diameter = max(self.max_diameter, leftHeight + rightHeight)
                print("max diam ", self.max_diameter)
                return height_node
        
        dfs(root)

        return self.max_diameter